src/main/java/com/redhat/pantheon/model/api/SlingResourceAdapterFactory.java

```diff
@@ -0,0 +1,65 @@
+package com.redhat.pantheon.model.api;
+
+import org.apache.sling.api.adapter.AdapterFactory;
+import org.apache.sling.api.resource.Resource;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+
+import static org.apache.sling.api.adapter.AdapterFactory.ADAPTABLE_CLASSES;
+import static org.apache.sling.api.adapter.AdapterFactory.ADAPTER_CLASSES;
+
+/**
+ * An adapter factory for model objects of type SlingResource. This component makes sure that
+ * calls to {@link Resource#adaptTo(Class)} with a subtype of {@link SlingResource} as a parameter
+ * will return an object.
+ *
+ * Types need to registered as adapter classes by simply adding a property in the @{@link Component}
+ * definition at the top of this class.
+ */
+@Component(
+        name = "Pantheon - JCRModel Adapter Factory",
+        service = AdapterFactory.class,
+        property = {
+                ADAPTABLE_CLASSES + "=org.apache.sling.api.resource.Resource",
+                ADAPTER_CLASSES + "=com.redhat.pantheon.model.Module",
```

Comment:
I tried `SlingResource` in hopes the adapter factory system would understand it, but alas it takes the specific types you want to adapt. You are correct in that we need to add an entry here for each concrete class that we want to automatically adapt from a Resource. It's just a line so I thought it might not be a big deal, but it is a prerequisite unfortunately.

As for the code you mention, that line comes from `SlingResource` I presume. That slightly different as I didn't necessarily want to adapt `SlingResource` objects to anything in particular, so I just invoke the underlying `Resource`'s `adaptTo` method. The reason for the adapter factory is the inverse adaptation: from `Resource` to one of our new models; that's the one we are plugging in to here.

Reply:
Ahhh. That makes perfect sense, thanks for clarifying. I forgot that `adaptTo` works both ways and erroneously assumed that these were related. That will introduce a small wrinkle when we try and create new Modules that don't have an underlying Resource, but that's a problem for another day.